File: services/ingest-node/src/middleware/rate_limit.js
"use strict";

/**
 * Simple in-memory sliding-window rate limiter per IP (non-Express).
 * Returns true if the request may proceed, false if rejected.
 */
const hits = new Map();

function checkRateLimit(req, res, env) {
  const ip = req.socket.remoteAddress || "unknown";
  const now = Date.now();
  const windowMs = env.RATE_LIMIT_WINDOW_MS;
  const max = env.RATE_LIMIT_MAX;

  let record = hits.get(ip);
  if (!record || now - record.start > windowMs) {
    record = { start: now, count: 0 };
    hits.set(ip, record);
  }

  record.count += 1;

  if (record.count > max) {
    res.writeHead(429, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ error: "rate_limited", message: "too many requests" }),
    );
    return false;
  }

  return true;
}

// Cleanup every 60s
setInterval(() => {
  const now = Date.now();
  for (const [ip, record] of hits) {
    if (now - record.start > 60000) hits.delete(ip);
  }
}, 60_000).unref();

module.exports = { checkRateLimit };

File: services/api-go/internal/middleware/ratelimit.go
package middleware

import (
	"net/http"
	"sync"
	"time"
)

// RateLimit provides a simple in-memory per-IP sliding window rate limiter.
func RateLimit(maxRequests int, window time.Duration) func(http.Handler) http.Handler {
	type record struct {
		count int
		start time.Time
	}

	var mu sync.Mutex
	hits := make(map[string]*record)

	// Background cleanup
	go func() {
		ticker := time.NewTicker(window)
		defer ticker.Stop()
		for range ticker.C {
			mu.Lock()
			now := time.Now()
			for ip, r := range hits {
				if now.Sub(r.start) > window {
					delete(hits, ip)
				}
			}
			mu.Unlock()
		}
	}()

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ip := r.RemoteAddr
			mu.Lock()
			rec, ok := hits[ip]
			now := time.Now()
			if !ok || now.Sub(rec.start) > window {
				rec = &record{start: now, count: 0}
				hits[ip] = rec
			}
			rec.count++
			count := rec.count
			mu.Unlock()

			if count > maxRequests {
				http.Error(w,
					`{"error":"rate_limited","message":"too many requests"}`,
					http.StatusTooManyRequests,
				)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}
